tests/usb_interfaces.test.cpp

```diff
@@ -0,0 +1,737 @@
+                                    to_le_bytes(p_wIndex),
+                                    to_le_bytes(p_wLength)));
+      return true;
+    }
+    return false;
+  }
+
+  operator std::span<u8 const>() const
+  {
+    return m_desc_arr;
+  }
+
+  constexpr byte& bInterfaceNumber()
+  {
+    return m_desc_arr[0];
+  }
+
+  constexpr byte& bAlternateSetting()
+  {
+    return m_desc_arr[1];
+  }
+
+  constexpr byte& bNumEndpoints()
+  {
+    return m_desc_arr[2];
+  }
+
+  constexpr byte& bInterfaceClass()
+  {
+    return m_desc_arr[3];
+  }
+
+  constexpr byte& bInterfaceSubClass()
+  {
+    return m_desc_arr[4];
+  }
+
+  constexpr byte& bInterfaceProtocol()
+  {
+    return m_desc_arr[5];
+  }
+
+  constexpr byte& iInterface()
+  {
+    return m_desc_arr[6];
+  }
+
+  mock_usb_endpoint m_generic_ep;
+  std::string_view m_name;
+
+  bool m_test_feature = false;          // feature 0
+  bool m_another_test_feature = false;  // feature 1
+
+private:
+  bool m_write_desc_has_ran = false;
+  std::array<byte, 7> m_desc_arr;
+};
+
+class mock_multi_interface : public usb_interface
+{
+public:
+  ~mock_multi_interface() override = default;
+
+  mock_multi_interface(std::string_view p_iface_name_one,
+                       std::string_view p_iface_name_two)
+    : m_name_one(p_iface_name_one)
+    , m_name_two(p_iface_name_two) {};
+
+  struct mock_iface_descriptor
+  {
+    byte num;
+    byte alt_settings;
+    byte str_idx;
+    bool feature;
+  };
+
+private:
+  [[nodiscard]] descriptor_delta driver_write_descriptors(
+    endpoint_writer const& p_callback,
+    u8 p_iface_number,
+    u8 p_str_starting_idx) override
+  {
+    std::array<byte const, 8> iad_buf{
+      iad_length,
+      iad_type,
+      0,                    // first interface
+      2,                    // iface count
+      0,                    // class
+      0,                    // subclass
+      0,                    // proto
+      p_str_starting_idx++  // string idx
+    };
+
+    std::array<byte const, 2> iface_header = { iface_desc_length,
+                                               iface_desc_type };
+    std::array<byte const, 5> static_desc_vars = {
+      0,  // altsettings
+      1,  // num endpoints
+      0,  // class
+      0,  // subclass
+      0,  // protocol
+    };
+
+    m_iface_one = { p_iface_number++, 0, p_str_starting_idx++, false };
+    m_iface_two = { p_iface_number++, 0, p_str_starting_idx++, false };
+
+    std::array<byte const, 1> iface_one_arr({ m_iface_one.num });
+    std::array<byte const, 1> iface_one_str_arr({ m_iface_one.str_idx });
+
+    std::array<byte const, 1> iface_two_arr({ m_iface_two.num });
+    std::array<byte const, 1> iface_two_str_arr({ m_iface_two.str_idx });
+
+    auto span_arr = make_scatter_bytes(iad_buf,
+
+                                       // iface one
+                                       iface_header,
+                                       std::span(iface_one_arr),
```

Comment:
```suggestion
                                       iface_one_arr,
```

You shouldn't have to wrap this in span. The make_scatter_bytes should do this automatically for you so long as the type is convertable. Anything with a .data and .size  and converting to bytes is compatible like your array or a vector.